### PR TITLE
fix(softmax): validate mask shape matches inputs shape

### LINKED ISSUE #19722: softmax sliently generate a wrong output when the mask has an  incompatible shape
As shown in this [colab](https://colab.research.google.com/drive/1B5c-9QEwn8GWvi0kOp_mPvSt5BGV34tM?usp=sharing), the mask can arbitrarily change the shape of output.
Although the document requires that the mask has the same shape with input, no check exists  actually.
In my project, the shape of input is hard to compute and this potential incompatibility can affect the result and is hard to debug since softmax now silently generates an wrong output.
Currently the output is:

You can identify the shape changed by mask(line 2 and line 3 of output), which is different from the correct shape computed by compute_output_shape() (line4).

### PR DESCRIPTION
**Pull request by rstar327**
## Summary
- Add shape validation in `Softmax.call()` when `mask` is provided
- Previously, an incompatible mask shape would silently broadcast via `numpy.where`, producing wrong output shape
- Now raises `ValueError` with a clear message when shapes don't match
- Adds test for mask shape mismatch

Fixes #19722

### CODE REVIEW

File: keras/src/layers/activations/softmax.py [hunk 1]
```diff
@@ -52,6 +52,12 @@ def __init__(self, axis=-1, **kwargs):
 
     def call(self, inputs, mask=None):
         if mask is not None:
+            if inputs.shape != mask.shape:
+                raise ValueError(
+                    "The `mask` must have the same shape as `inputs`. "
+                    f"Received: inputs.shape={inputs.shape}, "
+                    f"mask.shape={mask.shape}"
+                )
```
Comment (hertschuh) [inline keras/src/layers/activations/softmax.py]: Please look at this thread: https://github.com/keras-team/keras/pull/19736#discussion_r1607342013

The mask doesn't need to be the same shape as the inputs, it just needs to be broadcastable to the input shape.

So:
- please update the docstring above to mention the broadcast vs. the exact shape match
- the bug mentioned in https://github.com/keras-team/keras/issues/19722 is that the broadcast happens both ways, if mask is "bigger" than the inputs, the inputs are broadcast. This is weird and we should disallow that. But the criteria is more complicated.

**Review by hertschuh — COMMENTED** [review]
Thanks for working on this!

**Comment by rstar327** [comment]
Done, updated validation to allow broadcastable masks while disallowing masks larger than inputs.
Updated docstring and added tests for broadcastable, fewer-dim, mismatch, and too-many-dims cases.
